File: 23-merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<pair<int,ListNode*> , vector<pair<int,ListNode*>> , greater<pair<int,ListNode*>> >min_heap;
        
        for(auto list:lists){
            if(list){
                min_heap.push({list->val,list});
            }
        }

        ListNode* dummy=new ListNode(-1);
        ListNode* head=dummy;

        while(!min_heap.empty()){
            ListNode* curr=min_heap.top().second;
            min_heap.pop();
            head->next=curr;
            head=head->next;

            if(curr->next){
                min_heap.push({curr->next->val,curr->next});
            }
        }
        return dummy->next;
    }
};

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: ListNode | None, list2: ListNode | None) -> ListNode | None:
        
        if not list1:
            return list2

        if not list2:
            return list1

        if list1.val <= list2.val:
            list1.next=self.mergeTwoLists(list1.next,list2)
            return list1
        else:
            list2.next=self.mergeTwoLists(list1,list2.next)
            return list2
        
        return None

    def partitionAndMerge(self, s:int, e:int, lists:list[ListNode|None]):
        if s>e:
            return None
        if s==e:
            return lists[s]

        mid= s+ (e-s)//2
        l1=self.partitionAndMerge(s,mid,lists)
        l2=self.partitionAndMerge(mid+1,e,lists)

        return self.mergeTwoLists(l1,l2)
    def mergeKLists(self, lists: list[ListNode | None]) -> ListNode | None:
        return self.partitionAndMerge(0,len(lists)-1,lists)
        